### PR TITLE
batches: store key ID alongside credentials, take two

### PR DESCRIPTION
**Pull request by LawnGnome**
This is a funkier version of #20689. Like that PR, it stores the key ID next to the encrypted data for site and user credentials. Unlike that PR, it reunifies the `credential` and `credential_enc` fields into a single field, performing the required SQL hackery to make this work regardless of the previous state.

To finish making this align with other encryption users, I've extended the OOB migrations to support down migrations under the same conditions as they do. I'll update the documentation PR tomorrow once we've (hopefully) landed this.

### CODE REVIEW

**Review by mrnugget — APPROVED** [review]

**Comment by mrnugget** [comment]
Deferring to @eseliger for final approval, since I'm not familiar with that part of the codebase and am not comfortable on the decision to merge/not merge resting on my approval here 😄

File: enterprise/internal/batches/background/user_credential_migrator.go [hunk 1]
```diff
@@ -82,5 +92,49 @@ func (m *userCredentialMigrator) Up(ctx context.Context) error {
 }
 
 func (m *userCredentialMigrator) Down(ctx context.Context) error {
-	return errors.New("down migration is not supported for encrypting user credentials")
+	if !m.allowDecrypt {
+		log15.Warn("cannot run userCredentialMigrator.Down when decryption isn't allowed")
```
Comment (eseliger) [inline enterprise/internal/batches/background/user_credential_migrator.go:96]: If this threw an error, the migrations UI would display it. Not sure if we want that.

File: enterprise/internal/batches/store/site_credentials.go [hunk 2]
```diff
@@ -171,18 +180,28 @@ SELECT
...
 
 func listSiteCredentialsQuery(opts ListSiteCredentialsOpts) *sqlf.Query {
 	preds := []*sqlf.Query{sqlf.Sprintf("TRUE")}
-	if opts.OnlyUnencrypted {
-		preds = append(preds, sqlf.Sprintf("credential_enc IS NULL"))
+	if opts.RequiresMigration {
+		preds = append(preds, sqlf.Sprintf("encryption_key_id IN ('', %s)", btypes.SiteCredentialPlaceholderEncryptionKeyID))
+	}
+	if opts.OnlyEncrypted {
+		preds = append(preds, sqlf.Sprintf("encryption_key_id <> ''"))
```
Comment (eseliger) [inline enterprise/internal/batches/store/site_credentials.go:192]: wouldn't this need to filter out the ones that have the placeholder encryption key, too?

**Review by eseliger — APPROVED** [review]

File: enterprise/internal/batches/background/user_credential_migrator.go (hunk 1 above)
Comment (LawnGnome) [reply enterprise/internal/batches/background/user_credential_migrator.go:96]: The external account migrations don't return errors in this case, so I decided to imitate that here.

File: enterprise/internal/batches/store/site_credentials.go (hunk 2 above)
Comment (LawnGnome) [reply enterprise/internal/batches/store/site_credentials.go:192]: Nope! They're encrypted; we just don't know what key was used at the point the inline migration runs. So, in this case, we really do just want to check that it's not an empty string.

**Comment by LawnGnome** [comment]
I'm going to merge this just to unblock it and get it done, but if there's feedback next week I'm happy to come back to it.

**Pull request merged**